tests/test_read.py

```diff
@@ -319,3 +321,22 @@ def test_read_out_mask(path_rgb_byte_tif, out):
     with rasterio.open(path_rgb_byte_tif) as src:
         with pytest.raises(ValueError):
             src.read(indexes=[2], out=out)
+
+
+def test_chained_io_errors(path_rgb_byte_tif):
+    """Get chained exceptions."""
+    with rasterio.open("tests/data/corrupt.tif") as src:
+        with pytest.raises(RasterioIOError) as excinfo:
+            src.read()
+        assert isinstance(excinfo.value.__context__, CPLE_AppDefinedError)
+        assert excinfo.value.__context__.errmsg.startswith(
+            "tests/data/corrupt.tif, band 1: IReadBlock failed at X offset 1, Y offset 1"
+        )
+        assert isinstance(excinfo.value.__context__.__cause__, CPLE_AppDefinedError)
+        assert isinstance(
+            excinfo.value.__context__.__cause__.__cause__, CPLE_AppDefinedError
+        )
+        assert (
+            excinfo.value.__context__.__cause__.__cause__.errmsg
+            == "TIFFFillTile:Read error at row 512, col 0, tile 3; got 38232 bytes, expected 47086"
+        )
```

Comment:
A new test which asserts that the existing behavior on trying to read imagery from a busted GeoTIFF is preserved, and also that we can, by traversing to context and cause, get the root error.

Reply:
corrupt.tif is a copy of cogeo.tif, but with the following script applied to it

```python
from tifffile import TiffFile

with TiffFile("corrupt.tif", mode='r+b') as tif:
    tif.pages[0].tags["TileOffsets"].overwrite((131731, 169310, 207930, 260000))  # 250938))
```

Moving the offset of the last tile causes libtiff to fail when trying to read it.